provider/core/src/refcount.rs

```diff
@@ -0,0 +1,144 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+use core::mem::size_of;
+use core::{mem::ManuallyDrop, ptr::NonNull};
+use alloc::sync::Arc;
+use alloc::{string::String, rc::Rc};
+use alloc::boxed::Box;
+use stable_deref_trait::{StableDeref, CloneStableDeref};
+
+type WrappedType = Option<Box<[u8]>>;
+
+// Safe as a sentinel because 0x1 is not a valid usize address
+const SENTINEL: *const WrappedType = 1usize as *const WrappedType;
+
+/// A thin wrapper over [`NonNull`] that works with `*const T`
```

Comment:
why don't we just work with `*mut T`?

Reply:
because `Rc::into_raw` and `Rc::from_raw` are in `*const T`

I merged all this stuff into a single type with the `SelectedRc` import cfg so now this code is a lot cleaner and we don't need this wrapper abstraction.